drivers/dma/dma_rza2.c

```diff
@@ -0,0 +1,1541 @@
+}
+
+static int dma_get_data_size(uint32_t data_size)
+{
+	switch (data_size) {
+	case 1:
+		return 0;
+	case 2:
+		return 1;
+	case 4:
+		return 2;
+	case 8:
+		return 3;
+	case 16:
+		return 4;
+	case 32:
+		return 5;
+	case 64:
+		return 6;
+	case 128:
+		return 7;
+	default:
+		return -EINVAL;
+	}
+}
+
+static int get_free_and_set_unlocked(uint32_t *descrs_busy)
+{
+	int chunk;
+
+	chunk = find_lsb_set(~(*descrs_busy));
+	if (chunk && chunk <= CONFIG_DMA_RZA2_DESCRS_CHUNKS) {
+		*descrs_busy |= BIT(chunk - 1);
+		return chunk - 1;
+	}
+
+	return -1;
+}
+
+static int free_chunk(const struct device *dev, int chunk)
+{
+	struct dma_rza2_data *data = dev->data;
+	k_spinlock_key_t key;
+
+	if (chunk >= CONFIG_DMA_RZA2_DESCRS_CHUNKS || chunk < 0) {
+		return -EINVAL;
+	}
+	key = k_spin_lock(&data->lock);
+	data->descrs_busy &= ~BIT(chunk);
+	k_spin_unlock(&data->lock, key);
+
+	return 0;
+}
+
+static int get_descrs(const struct device *dev, struct rza2_dma_link_descriptor **descr)
+{
+	k_spinlock_key_t key;
+	struct dma_rza2_data *data = dev->data;
+	int chunk;
+
+	if (!descr) {
+		return -EINVAL;
+	}
+
+	key = k_spin_lock(&data->lock);
+
+	chunk = get_free_and_set_unlocked(&data->descrs_busy);
+	if (chunk < 0) {
+		*descr = NULL;
+	} else {
+		*descr = &data->descr_pool[chunk * CONFIG_DMA_RZA2_MAX_DESCRS];
+	}
+
+	k_spin_unlock(&data->lock, key);
+
+	return chunk;
+}
+
+static bool not_aligned(const struct device *dev, uint32_t addr)
+{
+	const struct dma_rza2_config *cfg = dev->config;
+
+	if (cfg->addr_alignment == 0) {
+		return false;
+	}
+
+	return !!(addr & (cfg->addr_alignment - 1));
+}
+
+static int rza2_construct_link_chain(const struct device *dev, struct dma_config *dma_cfg,
+				     struct dma_rza2_channel *chan, uint32_t ch_cfg)
+{
+	int i;
+	struct dma_block_config *block = dma_cfg->head_block;
+	uint32_t total_bytes = 0;
+
+	__ASSERT(chan, "channel should be provided");
+
+	if (!block) {
+		LOG_ERR("no dma blocks were provided");
+		return -EINVAL;
+	}
+
+	chan->chunk = get_descrs(dev, &chan->descrs);
+	if (chan->chunk < 0) {
+		LOG_ERR("%s: unable to allocate descriptor list for dma transfer", __func__);
+		return -ENOMEM;
+	}
+
+	for (i = 0; i < dma_cfg->block_count; i++) {
+		/* Check source_address and dest_address alignment */
+		if ((not_aligned(dev, block->dest_address)) ||
+		    (not_aligned(dev, block->source_address))) {
+			LOG_ERR("%s: buffers are not properly aligned", __func__);
+			free_chunk(dev, chan->chunk);
+			return -EINVAL;
+		}
+
+		chan->descrs[i].src_addr = Z_MEM_PHYS_ADDR(block->source_address);
+		chan->descrs[i].dest_addr = Z_MEM_PHYS_ADDR(block->dest_address);
+		chan->descrs[i].trans_byte = block->block_size;
+		chan->descrs[i].interval = 0;
+		chan->descrs[i].config = ch_cfg;
+		chan->descrs[i].header = HDR_LV;
+
+		/* DMA controller supports only 1 interval so apply the bigger one which enabled*/
+		if (block->source_gather_en) {
+			chan->descrs[i].interval = block->source_gather_interval;
```

Comment:
looks like `source_gather_interval` and `dest_scatter_interval` aren't timeout intervals, they are address intervals